### ISSUE TITLE
ZMQStream queue

### ISSUE THREAD

**Issue by a1ezzz** [zeromq/pyzmq#1095]
Is there a way to gracefully discard messages that were flushed by ZMQStream?

In details. I have a client (REQ) that sends messages to a server (REP). The client and the server are ZMQStream instances. I want the client to drop messages, that were not delivered in a specific timeout. I've tried many zmq flags (like SNDTIMEO, SNDHWM, BACKLOG, HANDSHAKE_IVL, RECONNECT_IVL, RECONNECT_IVL_MAX, IMMEDIATE, LINGER) to resolves this issue without success. All the messages are stored inside ZMQStream queue till server goes up. 

In the following code example messages are postponed till connection is made: https://gist.github.com/a1ezzz/ad22dfa2e5a8e1e89077a9eecbbfcd03#file-zmqstream-py

**Comment by minrk** [comment]
I don't believe there is a mechanism to discard unsent messages on ZMQStream, but I think SNDTIMEO should work if you use the newer `zmq.eventloop.future.Socket`

**Comment by a1ezzz** [comment]
Thank you for the answer. But I'm still confused since SNDTIMEO doesn't do anything useful.

I've updated the code example to use zmq.eventloop.future.Context (and so on zmq.eventloop.future.Socket), and the result is the same. I`ve tried to use a different combination of flags: SNDTIMEO, IMMEDIATE, LINGER and the most interesting situation is with IMMEDIATE flag. With this flag, messages won't be sent right after the new connection, but they aren't discarded - they will be sent after new zmq.eventloop.future.Socket.send method call.

It looks like these messages are cached inside ZMQStream (ZMQStream._send_queue queue I suppose). And looks like they do not interact with any context or socket until a connection is made.
